daemon/daemon.go

```diff
@@ -222,14 +222,21 @@ func (d *OrderedDaemon) waitGroupForLastPriority() *sync.WaitGroup {
 }
 
 func (d *OrderedDaemon) shutdown() {
-	d.lock.Lock()
-	defer d.lock.Unlock()
-
 	d.stopped.Set()
```

Comment:
```suggestion
	if !d.stopped.SetToIf(false, true) {
		return
	}
```

In the current version you cannot shutdown twice. (It would lock and then `running` will be `false`.) Now, a double shutdown could happen. Although this is strongly discouraged, it would still be good to fix.
By atomically setting and checking `stopped` this should be fixed.

Consider adding a test for this case. Something like the following should work:
```go
d := daemon.New()
err := d.BackgroundWorker("A", func(shutdownSignal <-chan struct{}) {
	<-shutdownSignal
	// sleep longer than the grace time before shutting down
	time.Sleep(2 * graceTime)
})
require.NoError(t, err)

d.Start()
time.Sleep(graceTime)

d.Shutdown()
time.Sleep(graceTime)
d.ShutdownAndWait()
```
This should hopefully fail before and pass with the suggestion :stuck_out_tongue:

Reply:
I just realize that with this version: the second ShutdownAndWait call never blocks even though we are still not completely shut down.
So an alternative would be to introduce `stoppedOnce sync.Once` instead of `stopped`. This would lead to a block also for the second call.

Reply:
Since this is an edge-case anyways. I don't really mind which option you want go for @lzpap

Reply:
Good suggestion! Indeed the test fails before with panic because the second shutdown is trying to close the already closed `shutdownSignal` channel of the workers.

Reply:
@Wollac okay, let me see, I didn't notice your new comments :D

Reply:
> I just realize that with this version: the second ShutdownAndWait call never blocks even though we are still not completely shut down.
> So an alternative would be to introduce `stoppedOnce sync.Once` instead of `stopped`. This would lead to a block also for the second call.

I'm a bit confused about the term `block` here.
The desired behavior here is to prevent calling a second shutdown while one is in progress, right? So one option is to just return from `shutdown()` if we see that the daemon is already stopped. Is this what you mean by blocking?

Currently, `stopped` is only set to `true` in `shutdown()`, so with this conditional set you suggested we essentially reach the same `do it once` behavior, no?

Reply:
Yes, that is the key purpose. However there is a subtle difference in the behavior:
Let's say you call `Shutdown` and then, while we are still trying to stop everything, `ShutdownAndWait` (just like in the test).
Now the question is, what is the expected behavior of `ShutdownAndWait`:
- Is it supposed to _only_ return once everything has been successfully stopped
- or should it just cause no panic?

When using `sync.Once` we get the behavior that `ShutdownAndWait` only returns when a shutdown call has been successful. I.e. when called at the same time, `ShutdownAndWait` will block until the first shutdown call is done. This means that as a user, when  `ShutdownAndWait` returns we are 100% sure that everything is shut down.
Without blocking, it could still be the case that some other process is still busy shutting down so we have no guarantee and would need to check `IsRunning` to be 100%.

To see this you could add  `assert.False(t, d.IsRunning())` after the final `d.ShutdownAndWait()` in the test and it would fail.

I guess the "blocking" solution is a wee bit cleaner, but it would also perfectly fine with me to assume that `Shutdown()` is not concurrency safe and if it is called from multiple go routines the user needs to know what they are doing